sweep/fee_bumper.go

```diff
@@ -197,3 +213,460 @@ func (b *BumpResult) Validate() error {
+		quit:            make(chan struct{}),
+	}
+}
+
+// Broadcast is used to publish the tx created from the given inputs. It will,
+// 1. init a fee function based on the given strategy.
+// 2. create an RBF-compliant tx and monitor it for confirmation.
+// 3. notify the initial broadcast result back to the caller.
+// The initial broadcast is guaranteed to be RBF-compliant unless the budget
+// specified cannot cover the fee.
+//
+// NOTE: part of the Bumper interface.
+func (t *TxPublisher) Broadcast(req *BumpRequest) (<-chan *BumpResult, error) {
+	log.Tracef("Received broadcast request: %s", newLogClosure(
+		func() string {
+			return spew.Sdump(req)
+		})())
+
+	// Attempt an initial broadcast which is guaranteed to comply with the
+	// RBF rules.
+	result, err := t.initialBroadcast(req)
+	if err != nil {
+		log.Errorf("Initial broadcast failed: %v", err)
+
+		return nil, err
+	}
+
+	// Create a chan to send the result to the caller.
+	subscriber := make(chan *BumpResult, 1)
+	t.subscriberChans.Store(result.requestID, subscriber)
+
+	// Send the initial broadcast result to the caller.
+	t.handleResult(result)
+
+	return subscriber, nil
+}
+
+// initialBroadcast initializes a fee function, creates an RBF-compliant tx and
+// broadcasts it.
+func (t *TxPublisher) initialBroadcast(req *BumpRequest) (*BumpResult, error) {
+	// Create a fee bumping algorithm to be used for future RBF.
+	feeAlgo, err := t.initializeFeeFunction(req)
+	if err != nil {
+		return nil, fmt.Errorf("init fee function: %w", err)
+	}
+
+	// Create the initial tx to be broadcasted. This tx is guaranteed to
+	// comply with the RBF restrictions.
+	requestID, err := t.createRBFCompliantTx(req, feeAlgo)
+	if err != nil {
+		return nil, fmt.Errorf("create RBF-compliant tx: %w", err)
+	}
+
+	// Broadcast the tx and return the monitored record.
+	result, err := t.broadcast(requestID)
+	if err != nil {
+		return nil, fmt.Errorf("broadcast sweep tx: %w", err)
+	}
+
+	return result, nil
+}
+
+// initializeFeeFunction initializes a fee function to be used for this request
+// for future fee bumping.
+func (t *TxPublisher) initializeFeeFunction(
+	req *BumpRequest) (FeeFunction, error) {
+
+	// Get the max allowed feerate.
+	maxFeeRateAllowed, err := req.MaxFeeRateAllowed()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the initial conf target.
+	confTarget := calcCurrentConfTarget(t.currentHeight, req.DeadlineHeight)
+
+	// Initialize the fee function and return it.
+	//
+	// TODO(yy): return based on differet req.Strategy?
+	return NewLinearFeeFunction(
+		maxFeeRateAllowed, confTarget, t.cfg.Estimator,
+	)
+}
+
+// createRBFCompliantTx creates a tx that is compliant with RBF rules. It does
+// so by creating a tx, validate it using `TestMempoolAccept`, and bump its fee
+// and redo the process until the tx is valid, or return an error when non-RBF
+// related errors occur or the budget has been used up.
+func (t *TxPublisher) createRBFCompliantTx(req *BumpRequest,
```

Comment:
Had a discussion with @morehouse , and wanna hear more opinions here - do we think this fee climbing is necessary here as it may use more fees than necessary? cc @ProofOfKeags @Roasbeef @ziggie1984

Reply:
The only way I see this using more fees than necessary is if the step size is too large. If we use a small step-size I fail to see how we can overpay with this algorithm.